test/provider/expect.unittest.ts

```diff
@@ -0,0 +1,333 @@
+const REGISTRATION_TIMEOUT = 5000;
+const PENDING_TIMEOUT = 100;
+
+let model: Model;
+
+let mockAppDirectory: jest.Mocked<AppDirectory>;
+let mockEnvironment: jest.Mocked<Environment>;
+let mockApiHandler: jest.Mocked<APIHandler<APIFromClientTopic>>;
+
+beforeEach(async () => {
+    mockAppDirectory = new AppDirectory(null!) as jest.Mocked<AppDirectory>;
+    mockEnvironment = createMockEnvironmnent();
+    mockApiHandler = new APIHandler<APIFromClientTopic>() as jest.Mocked<APIHandler<APIFromClientTopic>>;
+    (mockApiHandler as PartiallyWritable<typeof mockApiHandler, 'onConnection'>).onConnection = new Signal<[Identity]>();
+    (mockApiHandler as PartiallyWritable<typeof mockApiHandler, 'onDisconnection'>).onDisconnection = new Signal<[Identity]>();
+
+    model = new Model(mockAppDirectory, mockEnvironment, mockApiHandler);
+
+    jest.resetAllMocks();
+    useFakeTime();
+});
+
+describe('When creating a directory FDC3 app', () => {
+    const testWindow: TestWindow = {
+        pendingTime: 990,
+        createdTime: 1000,
+        appType: 'directory'
+    };
+
+    describe('When the window is registered quickly', () => {
+        expectTest(testWindow, 3000, [
+            [
+                'When a window is expected long before it is created, the window promise rejects',
+                {callTime: 500, finalizeTime: 500 + PENDING_TIMEOUT, result: 'reject-timeout'}
+            ],
+            [
+                'When a window is expected shortly before it is created, the window promise resolves',
+                {callTime: 950, finalizeTime: 3000, result: 'resolve'}
+            ],
+            [
+                'When a window is expected shortly after it is created but before it is registered, the window promise resolves',
+                {callTime: 1500, finalizeTime: 3000, result: 'resolve'}
+            ]
+        ]);
+    });
+
+    describe('When the window registration is delayed due to a slow app directory', () => {
+        expectTest(testWindow, 8000, [
+            [
+                'When a window is expected while the window is being registered, the window promise rejects',
+                {callTime: 1500, finalizeTime: 990 + REGISTRATION_TIMEOUT, result: 'reject-timeout'}
+            ],
+            [
+                'When a window is expected shortly before the window is registered, the window promise rejects',
+                {callTime: 7500, finalizeTime: 7500, result: 'reject-timeout'}
+            ]
+        ]);
+    });
+
+    describe('When a window is closed while pending', () => {
+        expectTest(
+            {
+                pendingTime: 990,
+                closeTime: 1000,
+                appType: 'directory'
+            },
+            3000,
+            [
+                [
+                    'When a window is expected shortly before it is created, the promise rejects',
+                    {callTime: 950, finalizeTime: 1000, result: 'reject-closed'}
+                ]
+            ]
```

Comment:
Whilst the block above are pretty easy to follow, the tuple here is much more obfuscated.

Should these be objects, rather than tuples? So the keys of the object make it easier to see the test setup?

Reply:
The difference between this block and the above blocks with the `TestWindow` not being declared inline with the `expectTest` call. I've fixed that, so things are consistent. Does that simplify things? The idea with this being a tuple is sticking to jest conventions for parameterized tests